File: 12. Flip Horizontally and Vertically.py
"""Topic: Flip Horizontally and Vertically"""
def flip_horizontal(A):
    return [list(reversed(r)) for r in A]
def flip_vertical(A):
    return list(reversed(A))

if __name__=='__main__':
    A=[[1,2,3],[4,5,6],[7,8,9]]
    print('Horizontal:')
    for r in flip_horizontal(A): print(' '.join(map(str,r)))
    print('Vertical:')
    for r in flip_vertical(A): print(' '.join(map(str,r)))

# TC: O(n*m)
# SC: O(n*m)

File: 12. Flip Horizontally and Vertically.cpp
// Topic: Flip Horizontally and Vertically
#include <bits/stdc++.h>
using namespace std;
int main()
{
    vector<vector<int>> A = {{1, 2, 3}, {4, 5, 6}, {7, 8, 9}};
    // horizontal flip (reverse each row)
    auto H = A;
    for (auto& r : H) reverse(r.begin(), r.end());
    // vertical flip (reverse row order)
    auto V = A;
    reverse(V.begin(), V.end());
    cout << "Horizontal:\n";
    for (auto& r : H)
    {
        for (auto v : r) cout << v << " ";
        cout << "\n";
    }
    cout << "Vertical:\n";
    for (auto& r : V)
    {
        for (auto v : r) cout << v << " ";
        cout << "\n";
    }
}

// TC: O(n*m)
// SC: O(n*m)
